src/JsonSchema/Validator.php

```diff
@@ -43,13 +41,50 @@ public function check($value, $schema = null, JsonPointer $path = null, $i = nul
      * types defined in the schema whenever possible. Note that the first argument is passed by reference,
      * so you must pass in a variable.
      *
+     * This method is now a simple alias, and exists only so that the API is not fundamentally broken for
+     * people who use this. This method should probably be officially deprecated at some point.
+     *
      * {@inheritDoc}
      */
-    public function coerce(&$value, $schema = null, JsonPointer $path = null, $i = null)
+    public function coerce(&$value, $schema = null)
     {
-        $validator = $this->factory->createInstanceFor('schema');
-        $validator->coerce($value, $schema);
+        // Preserve the current checkMode, because users of this API expect to be
+        // able to set it via the factory before calling this method.
+        $checkMode = $this->factory->getCheckMode() | self::CHECK_MODE_COERCE;
 
-        $this->addErrors(array_unique($validator->getErrors(), SORT_REGULAR));
+        return $this->validate($value, $schema, $checkMode);
+    }
+
+    /**
+     * Validates the given data against the schema and returns an object containing the results
+     * Both the php object and the schema are supposed to be a result of a json_decode call.
+     * The validation works as defined by the schema proposal in http://json-schema.org.
+     *
+     * @param mixed $value Decoded JSON document to validate
+     * @param mixed $schema Decoded JSON schema to validate against
+     * @param int $checkMode Set various options to change the validation behavior
+     *
+     * {@inheritDoc}
+     */
+    public function validate(&$value, $schema = null, $checkMode = self::CHECK_MODE_NORMAL)
```

Comment:
The `validate()` method is *supposed* to be the main entry point for everything except pass-by-value calls, which use `validateImmutable()` - so not sure what you're disagreeing with here? `check()` and `coerce()` are intended to be deprecated; the only reason they're still around is to avoid breaking things for people who are currently using them.

I would prefer to avoid conflating `validate()` with `_check()` - they're intended to be different things. Having the public API be the same thing as all the internal stuff seems rather messy IMO - that's what happened with `coerce()`, and half the point of this PR is to remove any need for the 'pass-everything-down-the-stack' approach. Unless you're suggesting we make every part of this library thread-safe, then going back to passing stuff around seems like a backwards step.